File: Structural/Adapter/Adaptor.py
from __future__ import annotations


class Target:
    def __init__(self) -> None:
        pass

    def request(self) -> str:
        return f"Default Behavior."


class Adapted:
    def __init__(self) -> None:
        pass

    def specific_request(self) -> str:
        return f"Behavior to adapt."


class Adaptor(Target):
    def __init__(self, adapted: Adapted) -> None:
        super(Adaptor, self).__init__()
        self.adapted = adapted

    def request(self) -> str:
        result = self.adapted.specific_request()

        return result


target = Target()
print(target.request())


adapted = Adapted()
print(adapted.specific_request())


adaptor = Adaptor(adapted)
print(adaptor.request())

File: Structural/Adapter/Adaptor.go
package main

import "fmt"

type target struct{}
type itarget interface {
	request() string
}

func newTarget() itarget {
	return &target{}
}

func (t *target) request() string {
	return "Default behavior"
}

type adapted struct{}
type iadapted interface {
	specificRequest() string
}

func newAdapted() iadapted {
	return &adapted{}
}

func (t *adapted) specificRequest() string {
	return "Specific behavior"
}

type Adaptor struct {
	target itarget
}

func (a *Adaptor) request(adapted iadapted) string {
	result := adapted.specificRequest()

	return result
}

func main() {
	target := newTarget()
	fmt.Println(target.request())

	adapted := newAdapted()
	fmt.Println(adapted.specificRequest())

	adaptor := &Adaptor{
		target: target,
	}

	adaptor.request(adapted)
}
